compiler/src/dotty/tools/dotc/transform/PureStats.scala

```diff
@@ -0,0 +1,30 @@
+package dotty.tools.dotc
+package transform
+
+import ast.{Trees, tpd}
+import core._, core.Decorators._
+import MegaPhase._
+import Types._, Contexts._, Flags._, DenotTransformers._
+import Symbols._, StdNames._, Trees._
+
+object PureStats {
+  val name: String = "pureStats"
+}
+
+/** Remove pure statements in blocks */
+class PureStats extends MiniPhase {
+
+  import tpd._
+
+  override def phaseName: String = PureStats.name
+
+  override def runsAfter: Set[String] = Set(Erasure.name)
+
+  override def transformBlock(tree: Block)(implicit ctx: Context): Tree =
+    val stats = tree.stats.flatMap {
+      case Typed(Block(stats, expr), _) if isPureExpr(expr) => stats
+      case stat => if !stat.symbol.isConstructor && isPureExpr(stat) then Nil else List(stat)
+    }
+    cpy.Block(tree)(stats, tree.expr)
```

Comment:
This implementation copies all blocks which leads in practice to large scale tree rewriting & allocation even if there are no pure statements. I'd try this instead:
```scala
val stats = Trees.flatten(
  tree.stats.mapConserve {
    case Typed(Block(stats, expr), _) if isPureExpr(expr) => Thicket(stats)
    case stat if !stat.symbol.isConstructor && isPureExpr(stat) => EmptyTree
    case stat => stat
  })
```
(flatten avoids tree copying if possible).

Reply:
That variant works well